### PR TITLE
E2E: re-add existing tests to the Jetpack group.

### PR DESCRIPTION
**Pull request by worldomonation**
#### Changes proposed in this Pull Request

This PR re-adds several existing suites of tests to the Jetpack group, to be run as part of the Jetpack E2E task in TeamCity.

Key changes:
1. removal of notion of paid/free themes from the library due to the deprecation of Premium Themes offering in [September 2021](pNEWy-e9X-p2).
2. install a [minor upgrade](https://github.com/eugene-manuilov/jest-runner-groups/issues/40) of the `jest-runner-groups` in order to return different test user account based on the group.
3. implement a new function `getTestAccountByJestGroup` that is a Jest Group-based counterpart to the existing `getTestAccountByFeature` function.

Details:
Change 1 is straightforward as the Premium Themes offering has been retired in September 2021, and so the method `ThemesPage.filterThemes` no longer make sense.

Change 2 and 3 are related aimed at supporting the use case seen in some specs that are run with different test users on an otherwise identical spec. 

In the Selenium framework, Jetpack tests were tagged with `@jetpack`, and a Calypso user for running these tests in the Jetpack context was provided as `jetpackUser` in the configuration file. 

As we now use `jest-runner-groups` the challenge I have run into is how to determine what group had been specified in the CLI. `jest-runner-groups@2.2.0` adds exports to the envvar of the groups that has been passed in, giving us the ability to use different test user depending on the outcome of something like this:

```
const username = !! process.env.JEST_GROUP_FEATUREA ? 'featureAUser' : 'featureBUser';
```

The function `getTestFeatureByJestGroup` is a simple multiple if statement to check whether a specific environment variable exists. Given the way we use `jest-runner-groups` by specifying only one group for a given run, this simple implementation should suffice.

#### Testing instructions

Ensure the following:
  - [ ] Gutenberg E2E (desktop)
  - [ ] Gutenberg E2E (mobile)
  - [ ] Calypso E2E (desktop)
  - [ ] Calypso E2E (mobile)
  - [ ] Pre-Release E2E
  - [ ] i18n E2E

Related to https://github.com/Automattic/wp-calypso/issues/61936

### CODE REVIEW

**Comment by worldomonation** [comment]
@dpasque I wanted to leave this as a draft for now to gauge feedback for a potentially contentious way I approached the test user switching mechanism, which does not play well with the `getTestAccountByFeature` function that you and Marcelo were instrumental in implementing.

I'll add more details as a comment on the piece of code that I'd like a second opinion on.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts
Comment (worldomonation) [inline packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: This approach is very simple because as noted in the PR description we currently use only one group per run in TeamCity. While it is possible that an engineer would want to run multiple jest group locally, resource constraints mean this is rather unlikely.

In other words, if we encounter a scenario in the future where a TeamCity build configuration may specify two Jest groups this function will only ever return one test account key.

Another issue with this implementation is that it adds yet another location where both the test account key and the Jest group name has to be updated should something change.

File: test/e2e/specs/media/media__upload.ts
Comment (worldomonation) [inline test/e2e/specs/media/media__upload.ts:37]: It was decided to simply add `Jetpack` to the list of parametrized test here for simplicity.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: I'm not 100% sold on this implementation myself and would like a second opinion on this. Originally, I had a more condensed `if/else` logic in the test file itself - similar to how we had the ternary for the `gutenbergEdgeUser` - and perhaps going back to that would be a better move?

Comment (dpasque) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: Pinging @fullofcaffeine so he can see and weigh in!

Comment (dpasque) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: Hmm, this is a tricky one!

While this is slick, I think my current inclination is that we should not couple the jest-runner-group and the user. While they can be related, I think they each serve a distinct purpose that might serve us well to keep separate. The way I see it...
- `jest-runner-groups` only answers the question of "which test files should be run together".
- We answer the question of "which user should we use" based on ENV variables and script specific variables.

Practically speaking, this means that we expose an ENV variable called `TARGET_JETPACK_SITES` and configure the jetpack TC builds to not only target the `jetpack` jest-runner-group but also to set `TARGET_JETPACK_SITES=1`.

In the future, if we ever had to step away from `jest-runner-groups`, I think the transition would be a lot more seamless if we didn't have to pull logic related to that from our scripts and utilities. 

Thoughts @worldomonation?

Comment (worldomonation) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: The point you raise about not coupling to a `jest-runner-groups` is super legit. As much as I don't really like it, setting our own environment variable in TeamCity could be the simplest and most maintainable solution here.

Comment (worldomonation) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: Made the changes in b7129a5603c62da5534497240679aaec332646a4. What do you think?

File: yarn.lock
Comment (worldomonation) [inline yarn.lock]: I decided to leave the installation of `jest-runner-group@2.2.0` as-is because we shouldn't be too far behind releases if we can.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts
Comment (dpasque) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: >Made the changes in https://github.com/Automattic/wp-calypso/commit/b7129a5603c62da5534497240679aaec332646a4. What do you think?

Yeah, I definitely prefer this solution personally, I think that looks really nice! Thanks for doing that! 😄 

If we're going to commit fully to that route, should we...
1. Add it to the `envVariables` (and all their fancy typing and default handling)
2. Make a `jetpack` [variant](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts#L13) and update the feature tables?

File: test/e2e/specs/media/media__edit.ts
Comment (fullofcaffeine) [inline test/e2e/specs/media/media__edit.ts:28]: I think this is fine for now, and is simple to scan and understand. However, in the long run, I'd say we should get rid of the `describe.each` iterations and use `getTestAccountByFeature` instead, which would allow us to more easily support other combinations that might help catch more bugs ahead of time (i.e it might even make sense to eventually run this as part of the GB edge tests as GB might indirectly affect this feature, too).

By the way, what kind of site does the `jetpackUser` hold? Is it a simple or AT site? Just trying to understand if it's really its own `siteType` or if it might intersect with AT and simple, which would mean it might be better represented as a `variant` or some other new attribute in the `get-test-account-by-feature` module.

File: test/e2e/specs/media/media__upload.ts
Comment (fullofcaffeine) [reply test/e2e/specs/media/media__upload.ts:37]: Ditto for https://github.com/Automattic/wp-calypso/pull/62324/files#r839027390.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts
Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: > jest-runner-groups only answers the question of "which test files should be run together".
> We answer the question of "which user should we use" based on ENV variables and script-specific variables.

That's how I understand it, too. I mean, at first sight, they could look like a feasible way to select accounts for tests, but they lack the flexibility of TC env params:
1. They're hardcoded in the shell script that's part of the TC build. You can't change them per build run like you can with `envs`
2. AFAIK, they're not ideal If we want to support more complex account combination criteria (i.e jetpack on Gutenberg edge) as we'd need to pass `@gutenberg-edge @jetpack` or `@gutenberg @jetpack @edge`. This would require us to make the actual build in TC more specific -- and we'd need to pass this to `jest` when calling it from the build shell script.

> > Made the changes in [b7129a5](https://github.com/Automattic/wp-calypso/commit/b7129a5603c62da5534497240679aaec332646a4). What do you think?
> 
> Yeah, I definitely prefer this solution personally, I think that looks really nice! Thanks for doing that! smile
> 
> If we're going to commit fully to that route, should we...
> 
>     1. Add it to the `envVariables` (and all their fancy typing and default handling)
> 
>     2. Make a `jetpack` [variant](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts#L13) and update the feature tables?

Looking at the way it's implemented now, It looks like the best way to represent `jetpack` is not by variant, but by a flag with the same name so that we're able to declare the following criteria:

```js
{jetpack: true, accountName: 'jetpackUser' }
{jetpack: false, accountName: 'defaultUser' }
```
In fact, we can't use `variant` as it's not mutually exclusive to Jetpack as it currently is used to define a `siteEditor`-specific test, and we could in theory have a jetpack site-editor test :), i.e:

```
{jetpack: true, accountName: 'defaultUser', variant: 'siteEditor' }
```

Assuming the Jetpack build only needs to work with this specific account/site and doesn't need to support any other combinations for now, we could call it `getTestAccountByFeature` like this:

```js
const accountName = getTestAccountByFeatureKey( { jetpack: !! envVariables.TARGET_JETPACK } )
```

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: **Actually**, I think another attribute would still be needed to "group" these criteria, otherwise, semantically, `{jetpack: false, accountName: 'defaultUser' }` would apply to anything else that's not jetpack, which is confusing and wrong. So, I guess it'd be best to have an additional optional `group` attribute and then define the rules as so:

```js
{jetpack: true, group: 'jetpack', accountName: 'jetpackUser' }
{jetpack: false, group: 'jetpack',  accountName: 'defaultUser' }
```

Then we'd call it as:

```js
const accountName = getTestAccountByFeatureKey( { jetpack: !! envVariables.TARGET_JETPACK, group: 'jetpack' } );
```

However...

If we want to keep a common interface - which is one of the main goals of this abstraction - we shouldn't be passing the `FeatureKey` directly to the `getTestAccoutnByFeatureKey` function. We'd do the following, instead:

1. Set a new env var (`TARGET_JETPACK`), which you already did;
2. Add this var to the corresponding `envVariables` type:

```js
export interface SupportedEnvVariables extends EnvVariables {
        // (...)
        TARGET_JETPACK: boolean;
}

const defaultEnvVariables: SupportedEnvVariables = {
      // (...)
      TARGET_JETPACK: false
};
```

3. Add the mapping in the [envToFeatureKey](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/utils/get-test-account-by-feature.ts) function:

```js
export function envToFeatureKey( envVariables: TestAccountEnvVariables ): FeatureKey {
	return {
		coblocks: envVariables.COBLOCKS_EDGE ? 'edge' : undefined,
		gutenberg: envVariables.GUTENBERG_EDGE ? 'edge' : 'stable',
		siteType: envVariables.TEST_ON_ATOMIC ? 'atomic' : 'simple',
                jetpack: envVariables.TARGET_JETPACK
	};
}
```

The criteria would look like this (to be defined [here](https://github.com/Automattic/wp-calypso/blob/trunk/packages/calypso-e2e/src/lib/utils/criteria-for-test-accounts.ts)):

```js
{coblocks: '*', gutengerg: '*', siteType: '*', jetpack: true, group: 'jetpack', accountName: 'jetpackUser'}
{coblocks: '*', gutengerg: '*', siteType: '*', jetpack: false, group: 'jetpack', accountName: 'defaultUser'}
```

Notice that for this to work (and have a better DX), we'd need some kind of wildcard support. I'm working on it as part of [this](https://github.com/Automattic/wp-calypso/pull/62235) draft PR.

Now we can just call it like:

```js
const accountName = getTestAccountNameByFeature( {group: 'jetpack', ...envToFeatureKey( envVariables ) } );
```

I'm not entirely sure about the `group` attribute. It seems redundant to have a `group: "jetpack"` + a `jetpack: boolean` attributes, but if we want to support defining the default account as part of the criteria for these group of tests, then we need a way to group and query by this group. 

Looking forward to reading your thoughts! @dpasque @worldomonation

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: By the way, I'm not suggesting you do this as part of this PR. I'm totally fine with the solution in https://github.com/Automattic/wp-calypso/commit/b7129a5603c62da5534497240679aaec332646a4 for now :)

File: test/e2e/specs/media/media__edit.ts
Comment (worldomonation) [reply test/e2e/specs/media/media__edit.ts:28]: @fullofcaffeine 

> I'd say we should get rid of the describe.each iterations and use getTestAccountByFeature instead

Yes, I do agree with this. If going down this path, we should add `@gutenberg` group to this spec file at the same time as the switch, which is what you also suggest later in the paragraph if I understand correctly.

> what kind of site does the jetpackUser hold?

I am confused by Jetpack things myself - I am not too familiar with Jetpack and the way it interacts with WordPress.com sites is really confusing. I wish there was documentation that exists (there may be that I am not aware of) which explains these things.

I checked Store Admin and didn't find the test site listed under anything that has an upgrade, so I would surmise it's a normal Simple site.

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: I did read it over a couple of times and my overarching priority is to not introduce any more complexity to the matrix we have going in `getTestAccountByFeature`, which is already quite complex from the looks of it. Therefore I want to avoid adding yet another variable and making everything more complicated than it already is.

To that end, since Jetpack tests are rather few and is not a priority for the Jetpack team - I think we can leave out the inclusion of `jetpack` feature type from the account matrix. 

I propose we continue witha simpler solution I have in https://github.com/Automattic/wp-calypso/commit/b7129a5603c62da5534497240679aaec332646a4 and make it abundantly clear that use of `TARGET_JETPACK` environment variable is an exception rather than the norm.

P.S. I need to read up on `getTestAccountByFeature` and try to understand the logic. Haven't been too involved in the code review related to this and my understanding is a bit weak.

Comment (dpasque) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: Hrm, this one is a doozy... I do feel like the jetpack inclusion kind of muddies up the feature criteria to account table 😕 . It feels like we might be making that table way more complicated to handle something that has very narrow application.

I think I'm currently landing on...
1. We should still add `TARGET_JETPACK` to the `envVariables` to get all the good default behavior and typing.
2. We can avoid adding it to the feature tables for now. Instead, in the few scripts that need this, we can do something like...

```typescript
let accountName;

if ( envVariables.TARGET_JETPACK) {
    accountName = 'jetpackUser'
} else {
    const features = envToFeatureKey( envVariables );
    accountName = getTestAccountByFeature( features );
}
```

Comment (fullofcaffeine) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: > It feels like we might be making that table way more complicated to handle something that has very narrow application.

> I propose we continue witha simpler solution I have in https://github.com/Automattic/wp-calypso/commit/b7129a5603c62da5534497240679aaec332646a4 and make it abundantly clear that use of TARGET_JETPACK environment variable is an exception rather than the norm.

That makes a lot of sense! We can always revisit this later if we need more test site [combos](https://github.com/Automattic/wp-calypso/pull/62324#discussion_r839027390) for JP. Thanks for sharing your thoughts!

File: test/e2e/specs/media/media__edit.ts
Comment (fullofcaffeine) [reply test/e2e/specs/media/media__edit.ts:28]: > Yes, I do agree with this. If going down this path, we should add @gutenberg group to this spec file at the same time as the switch, which is what you also suggest later in the paragraph if I understand correctly.

Yes, and I think in the long run it'd be nice to make a habit of using `getTestAccountByFeature` so that we can declare the accounts in a single place and give us much more flexibility to change/expand the accounts per criteria. This would move the parameterization out of the `describe.each` tables into TC builds (as env vars) that would then match the relevant entries in the central criteria table. However, as we discussed [here](https://github.com/Automattic/wp-calypso/pull/62324#discussion_r839787537), I think it's totally fine to postpone this given we're not really sure how JetPack fits into the criteria yet, and the current implementation is simple and works ;)

File: packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts
Comment (worldomonation) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]:
```
let accountName;

if ( envVariables.TARGET_JETPACK) {
    accountName = 'jetpackUser'
} else {
    const features = envToFeatureKey( envVariables );
    accountName = getTestAccountByFeature( features );
}
```

I will update the changes here to use this switching logic instead.

Comment (worldomonation) [reply packages/calypso-e2e/src/lib/utils/get-test-account-by-group.ts]: I've condensed the if/else to a ternary in the updated revision f4f8070f3d301422260c29624b926f2249e379e4.

File: test/e2e/specs/media/media__edit.ts
Comment (worldomonation) [reply test/e2e/specs/media/media__edit.ts:28]: I can try switching over the specs that still use `describe.each` to parametrize the tests to use `getTestAccountByFeature` in a subsequent PR, if you'd like!

**Comment by worldomonation** [comment]
Ready for re-review. The changes since last round of reviews are minor.

File: .teamcity/_self/projects/WPComTests.kt
Comment (dpasque) [inline .teamcity/_self/projects/WPComTests.kt]: Ack, two small things I missed here!

1. The name should match the value in `envVariables`, so `TEST_ON_JETPACK`.
2. It seems for booleans we actually use `"true"` or `"false"` instead of `1`/`0`.

**Review by dpasque — APPROVED** [review]
There's one more change to make to the TeamCity file that should happen before merge. But everything else looks great so approving! Thanks for doing this! 😄

File: .teamcity/_self/projects/WPComTests.kt
Comment (worldomonation) [reply .teamcity/_self/projects/WPComTests.kt]: Good catch, thanks!

Comment (worldomonation) [reply .teamcity/_self/projects/WPComTests.kt]: Changed to a checkbox type with default value of `true` for the Jetpack build.

**Pull request merged**
